transformer/edit.go

```diff
@@ -0,0 +1,104 @@
+/*
+ * skogul, edit transformer
+ *
+ * Copyright (c) 2019 Telenor Norge AS
+ * Author(s):
+ *  - Kristian Lyngstøl <kly@kly.no>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+package transformer
+
+import (
+	"fmt"
+	"log"
+	"regexp"
+	"sync"
+
+	"github.com/telenornms/skogul"
+)
+
+// Replace executes a set of regular expressions, in order
+type Replace struct {
+	Source      string `doc:"Metadata key to read from."`
+	Destination string `doc:"Metadata key to write to. Defaults to overwriting the source-key if left blank. Destination key will always be overwritten, e.g., even if the source key is missing, the key located at the destination will be removed."`
+	Regex       string `doc:"Regular expression to match."`
+	Replacement string `doc:"Replacement text. Can also use $1, $2, etc to reference sub-matches. Defaults to empty string - remove matching items."`
+	regex       *regexp.Regexp
+	once        sync.Once
+	err         error
+}
+
+// Transform executes the regular expression replacement
+func (replace *Replace) Transform(c *skogul.Container) error {
+	replace.once.Do(func() {
+		if replace.Destination == "" {
+			replace.Destination = replace.Source
+		}
+		replace.regex, replace.err = regexp.Compile(replace.Regex)
+	})
+	// Verify() should catch this, so there's no reasonable way this
+	// should happen. But in the off chance that a regex compiles on
+	// the first attempt but not the second.... (e.g.: some serious
+	// bugs)
```

Comment:
I don't think we need this. :p Especially if the verify should pick up a compile error of the regex, I think we should trust the underlying implementation enough to work consistently 😅

Reply:
But this assumes we run Verify() under all thinkable scenarios, though. I'm somehow less concerned about regexp failing to compile on the second run than us somehow failing to do the Verify() under some corner case in the future... And I figure asserts are never really wrong...

Admittedly, I wrote this BEFORE I wrote Verify(), but still....